### ISSUE TITLE
Ubo Picker broken on MSN.com

### ISSUE THREAD

**Issue by troysjanda** [uBlockOrigin/uBlock-issues#2346]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is not a support issue or a question. For any support, questions or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

using the picker block entire page and unable to select individual items. as seen in animated gif

![pickerIssuesOnMSN](https://user-images.githubusercontent.com/24624945/198747967-c0525ef2-d64d-472b-97a5-e42455ba92dc.gif)

sorry about the full screen was trying to be complete in the process

Debug info

- uBlock Origin: 1.44.5b21
- Firefox: 106
- filterset (summary): 
-   network: 273064
-   cosmetic: 127973
-   scriptlet: 21422
-   html: 825
- listset (total-discarded, last-updated): 
-   removed: 
-     plowe-0: null
-   added: 
-     adguard-spyware-url: 793-3, 4h.44m
-     block-lan: 44-0, 4h.44m
-     curben-phishing: 177631-9, 4h.39m
-     curben-pup: 192-0, 4h.43m
-     fanboy-annoyance: 81841-9490, 4h.43m
-     ublock-annoyances: 4733-9, 4h.43m
-   default: 
-     user-filters: 0-0, never
-     ublock-filters: 31834-58, 4h.44m
-     ublock-badware: 3947-0, 4h.44m
-     ublock-privacy: 288-0, 4h.44m
-     ublock-abuse: 76-0, 4h.44m
-     ublock-unbreak: 1886-3, 4h.44m
-     ublock-quick-fixes: 427-0, 4h.44m
-     easylist: 62415-337, 4h.44m
-     easyprivacy: 30822-1190, now
-     urlhaus-1: 11029-0, 4h.44m
- filterset (user): [empty]
- trustedset: 
-   added: [array of 44 redacted]
- hostRuleset: 
-   added: [array of 3 redacted]
- modifiedUserSettings: 
-   advancedUserEnabled: true
-   showIconBadge: false
-   webrtcIPAddressHidden: true
- modifiedHiddenSettings: 
-   autoCommentFilterTemplate: none
-   extensionUpdateForceReload: true
-   filterAuthorMode: true
-   filterOnHeaders: true
-   updateAssetBypassBrowserCache: true
-   userResourcesLocation: [redacted]
- supportStats: 
-   allReadyAfter: 366 ms (selfie)
-   maxAssetCacheWait: 518 ms

### A specific URL where the issue occurs.

MSN.com

### Steps to Reproduce

open msn.com
click the Ubo icon
select picker
attemp to pick item to block
whole page highlights

### Expected behavior

should be able to pick any item on page

### Actual behavior

Entire page is red and if you click it the picker windows opens and you only get 2 selection

-     ##entry-point
-     ##fluent-design-system-provider

if one is selected it blocks the entire page

### uBO version

Dev 1.44.5b21

### Browser name and version

Firefox 106.0.2

### Operating System and version

Windows 11 Enterprise 22621.755 22H2

**Comment by gorhill** [comment]
Everything under `entry-point` is inside a shadow root.

**Comment by troysjanda** [comment]
> Everything under `entry-point` is inside a shadow root.

Not sure what the means sir. thinking that it means that it can't be fixed?

**Comment by gorhill** [comment]
I don't know what can be done at the moment, this will need time for investigating, my first thought is to maybe have something like _"Block element in frame…"_ but except _"Block element in shadow…"_, if possible at all

**Comment by troysjanda** [comment]
ok, if you need any other information please ask. thanks for your quick replies.

**Comment by troysjanda** [comment]
>Well "Block element in shadow…" is probably not a good idea, that page uses shadows inside shadows inside shadows...

I personally don't use that page often but was trying to help someone on the reddit page and ran across it.

**Comment by troysjanda** [comment]
> I use a scriptlet for removing shadowroot nodes - https://github.com/uBlock-user/uBO-Scriptlets/blob/master/scriptlets.js#L3

I am already using your scriptlets in advance settings. is there a MSN.com filter I should be using?

**Comment by krystian3w** [comment]
IMO no possible good use rsre ("remove-shadowroot-elem") to safe use element picker.

**Comment by garry-ut99** [comment]
Duplicate of / invalid due to / related to https://github.com/uBlockOrigin/uBlock-issues/issues/2215 and https://github.com/uBlockOrigin/uBlock-issues/issues/803, wherein 2215  is already a duplicate of / invalid due to / related to 803.

**Comment by troysjanda** [comment]
> Also duplicate of / invalid due to / related to #2215 and #803 (wherein 2215 is already a duplicate of / invalid due to / related to 803).

This is not invalid as the picker is broken on the MSN site due to the shadow dom issues.  If  one is unable to use the picker to pick an element to hide, but instead gets a red overlay and unable to select anything that is broken. #2215 is about ads he had blocked showing again and at that time the response was unable to reproduce, and deemed invalid. #803 is a specific filter issue 

>Issue can be reproduced using #mainContent selector, so no need to get a EU IP address to reproduce.

As stated by Gorhill comment https://github.com/uBlockOrigin/uBlock-issues/issues/803#issuecomment-586712831

My issue is that the picker is broken when attempting to make selections on MSN site, due to ShadowDom.

**Comment by garry-ut99** [comment]
> as the picker is broken on the MSN site due to the shadow dom issues. If one is unable to use the picker to pick an element to hide, but instead gets a red overlay and unable to select anything that is broken.

Perfectly same like in 2215.

> https://github.com/uBlockOrigin/uBlock-issues/issues/2215 is about ads he had blocked showing again

A cherry picked the first part of the description, the other part ignored, despite the other part is the most important:

- 2215 : Description: areas I have blocked are now appearing and **the elemnt picker and element zapper are no tworking. When I try to use them the entire page is highlighted.**
- 2346 : Description: **using the picker block entire page and unable to select individual items. as seen in animated gif**
- https://github.com/uBlockOrigin/uAssets/issues/13052#issuecomment-1115795330 which was marked as duplicate of 803 Description: **But I can't block them, because the element picker can't pick it. the select element can't be picked.**

Duplicates: the same issues  (a broken picker which is highlighting the whole page) caused by the same cause (shadow DOM).

**Comment by krystian3w** [comment]
Hard but can be used: https://github.com/gorhill/uBlock/commit/52b46eb98bedfa746206e89ac016734cbbed092a I do not recommend yet for #2297 

anti-social demo in header (legacy 2024):

```adb
! works only with older DOM Tree/Nodes (2024 layout)
msn.com##fluent-design-system-provider > entry-point:shadow(div[class*="container"]) > me-stripe:shadow(msft-horizontal-card-slider) > me-stripe-tile:shadow(.me-stripe-tile-content) a[href*="facebook.com"]
! works only with older DOM Tree/Nodes  (2024 layout)
```

**Comment by fkoemep** [comment]
@krystian3w is that selector still experimental? I have a custom filter that always worked for enabling picture in picture when websites block it: 
```js 
*##video:remove-attr(disablePictureInPicture)
```

I'm realising that now it doesn't work on Disney+ because they hide it under a shadow-root under a custom element:
![image](https://github.com/uBlockOrigin/uBlock-issues/assets/32828909/55848fd2-cd65-463a-8fb3-df59392ccae0)


Any ideas how could I make it work?

**Comment by gorhill** [comment]
> Any ideas how could I make it work?

Try:

    [disney website hostname]##disney-web-player:shadow(video):remove-attr(disablePictureInPicture)

**Comment by fkoemep** [comment]
> > Any ideas how could I make it work?
> 
> Try:
> 
> ```
> [disney website hostname]##disney-web-player:shadow(video):remove-attr(disablePictureInPicture)
> ```

That worked! Had to modify "disablePictureInPicture" to lowercase, weird since I thought ``remove-attr`` is case insensitive.

I know this is probably bad practice but I generalized the rules to unblock PiP for every website, works for now:

```
*##video:remove-attr(disablePictureInPicture)
*##video:remove-attr(disablepictureinpicture)
*##*:shadow(video):remove-attr(disablePictureInPicture)
*##*:shadow(video):remove-attr(disablepictureinpicture)
```

**Comment by garry-ut99** [comment]
> fkoemep : That worked! Had to modify "disablePictureInPicture" to lowercase, weird since I thought `remove-attr` is case insensitive.

From [uBO remove-attr documentation](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#subjectremove-attrarg-subjectremove-classarg) : 

> arg: A plain string to match **exactly**, or a regex literal.

It seems exactly means really exactly, which means not only the whole attribute name from left to right, but also case sensivity too.

But on the other hand, you seem to be right too, from [w3c documentation](http://w3c.github.io/html-reference/documents.html#case-insensitivity): 

> [Attribute names](http://w3c.github.io/html-reference/syntax.html#attribute-name) for [HTML elements](http://w3c.github.io/html-reference/elements.html#html-elements) may be written with any mix of lowercase and uppercase letters that are a [case-insensitive match](http://w3c.github.io/html-reference/terminology.html#case-insensitive) for the names of the attributes given in the [HTML elements](http://w3c.github.io/html-reference/elements.html) section of this document; that is, attribute names are case-insensitive.

It has been fixed for [set-attr](https://github.com/gorhill/uBlock/wiki/Resources-Library#set-attrjs-): 
- https://github.com/uBlockOrigin/uBlock-issues/issues/3121

Anyway, as for now:
- you can still use regex with [ignoreCase](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/ignoreCase) flag `i`.
- also you can combine phrases in regex by using [Non-capturing group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Non-capturing_group)
- also you can combine CSS selectors

`*##video, audio:remove-attr(/disable(?:PictureInPicture|RemotePlayback)/i)`
`*##*:shadow(video):remove-attr(/disable(?:PictureInPicture|RemotePlayback)/i)`

**Comment by krystian3w** [comment]
Enough in Firefox use PiP button in URL then parameter no longer works anywhere (so only Chromium is susceptible).

**Comment by garry-ut99** [comment]
>  (so only Chromium is susceptible).

Or in Chromium browsers use one of many [PIP extensions](https://chromewebstore.google.com/search/pip?itemTypes=EXTENSION) (I didn't test them).

**Comment by krystian3w** [comment]
For now, this is not the possible (at least in 2024, there were no indications of this). Are many websites starting to use this to restore base64 ads or annoying placeholders?

**Comment by AlexJuarez** [comment]
When adding this feature where would tests go for this? And where would this new functionality live? Would somewhere near the current [elementFromPoint function in epicker.js](https://github.com/gorhill/uBlock/blob/bc5fa9fe9a1bcd7faee9694804624d3871122b7f/src/js/scriptlets/epicker.js#L919) be correct?

Here's the functionality I propose adding: A recursive grab for the next element in the shadowRoot (if present) "at" `x, y`. It returns the element with a top value closest to `y`.

```javascript
const getElementAtPoint = (x, y) => {
    const elem = document.elementFromPoint(x, y);
   
    const resolveElement = el => {
        if (el.shadowRoot) {
            let t = el.shadowRoot.elementFromPoint(x, y);
            if (t) {
                const { top } = t.getBoundingClientRect();
                if (top < y) {
                    return resolveElement(t);
                }
            }
        }
        return el;
    }
    
    return resolveElement(elem);
}
```

### Reference APIs:
- [shadowRoot](https://developer.mozilla.org/en-US/docs/Web/API/Element/shadowRoot)
- [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)
- [ShadowRoot.elementFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/elementFromPoint)

### Note:
It looks like shadowRoot.elementFromPoint isn't available if `pointer-events: none;` is set on the element. 
Maybe add a check as it finds each element and set the pointerEvents for that element, then unset when done.

```javascript
const elements = [];
if (window.getComputedStyle(elem)["pointer-events"] === "none") {
    elem.style.pointerEvents = "auto";
    elements.append(elem);
}
// do stuff
elements.forEach(el => el.style.pointerEvents = "none");
```

Thoughts?

**Comment by krystian3w** [comment]
Do you mean the situation where the picker won’t let you save when the root elements have `pointer-event: none`?
In uBOL, this has been fixed so that the button always works until there’s an actual syntax error, rather than a mismatch.

---
---

As for multiple nesting, nothing particularly interesting has been implemented yet (as far as I can see), as specifying a specific name doesn’t work and we don’t want a list like `:shadow(*):shadow(*):shadow(*)...`. I haven’t tested AdGuard’s [approach](https://github.com/AdguardTeam/Scriptlets/blob/master/wiki/about-scriptlets.md#hide-in-shadow-dom) (using scriptlets) to see how well it handles multiple nesting.

The actively developed tester ()https://ublockorigin.github.io/uBOL-home/tests/test-filters.html does not have a section under `:shadow(*)` to check if it works.
